File: bot/src/filters/hallucination-filter.ts
/**
 * ハルシネーションフィルター - Hallucination detection and filtering
 *
 * Whisper はまれに以下のようなハルシネーションを生成する:
 * - 同じフレーズの繰り返し: 「しょうがない しょうがない しょうがない」
 * - 意味不明な繰り返しパターン
 * - 特定の定型フレーズ（字幕提供、ご視聴ありがとうございましたなど）
 */
import { logger } from '../utils/logger.js';

export interface HallucinationFilterConfig {
  enabled: boolean;
  minRepetitionCount: number;
  maxRepetitionLength: number;
}

// 定型フレーズパターン（Whisper がよく生成するハルシネーション）
const HALLUCINATION_PATTERNS: RegExp[] = [
  // 典型的なハルシネーション（無音時に頻発）
  /^ありがとうございました[。.]?$/,
  /^ありがとうございます[。.]?$/,
  /ありがとうございました[。.]?$/,  // 文末のありがとうございましたも検出
  /^字幕提供.*$/,
  /^ご視聴ありがとうございました.*$/,
  /^チャンネル登録.*$/,
  /^お疲れ様でした[。.]?$/,
  /^最後までご視聴.*$/,
  /^ご清聴ありがとうございました.*$/,
  /^おやすみなさい[。.]?$/,
  /^では[、,]?また[。.]?$/,
  /^それでは[。.]?$/,
  /^以上です[。.]?$/,
  /^\.+$/, // ドットのみ
  /^,+$/, // カンマのみ
  /^[\s\u3000]+$/, // 空白のみ
  /(?:music|♪|♫)+/i, // 音楽記号
  /\[音楽\]/,
  /\[拍手\]/,
  /^\s*お\s*$/, // 単独の「お」
  /^\s*ん\s*$/, // 単独の「ん」
  /^はい[。.]?$/, // 単独の「はい」
  /^うん[。.]?$/, // 単独の「うん」
  /^えー[っと]*[。.]?$/, // えー、えーっと
];

export interface HallucinationFilterResult {
  text: string;
  wasFiltered: boolean;
  reason: string | null;
}

/**
 * ハルシネーションフィルター
 */
export class HallucinationFilter {
  private enabled: boolean;
  private minRepetitionCount: number;
  private maxRepetitionLength: number;
  private stats = {
    totalFiltered: 0,
    repetitionFiltered: 0,
    patternFiltered: 0,
  };

  constructor(config: Partial<HallucinationFilterConfig> = {}) {
    this.enabled = config.enabled ?? true;
    this.minRepetitionCount = config.minRepetitionCount ?? 3;
    this.maxRepetitionLength = config.maxRepetitionLength ?? 20;
  }

  /**
   * 繰り返しパターンを検出
   */
  detectRepetition(text: string): { isRepetition: boolean; phrase: string | null } {
    if (!text || text.length < 6) {
      return { isRepetition: false, phrase: null };
    }

    const t = text.trim();

    // スペースで分割して同じ単語の繰り返しを検出
    const words = t.split(/\s+/);
    if (words.length >= this.minRepetitionCount) {
      const uniqueWords = new Set(words);
      if (uniqueWords.size === 1) {
        return { isRepetition: true, phrase: words[0] };
      }
    }

    // サブストリングの繰り返しを検出
    for (
      let phraseLen = 2;
      phraseLen < Math.min(this.maxRepetitionLength + 1, Math.floor(t.length / 2) + 1);
      phraseLen++
    ) {
      const phrase = t.substring(0, phraseLen);
      let count = 0;
      let pos = 0;

      while ((pos = t.indexOf(phrase, pos)) !== -1) {
        count++;
        pos += phrase.length;
      }

      if (count >= this.minRepetitionCount) {
        const expectedLen = phrase.length * count;
        // 繰り返しがテキストの80%以上を占める
        if (expectedLen / t.length >= 0.8) {
          return { isRepetition: true, phrase };
        }
      }
    }

    return { isRepetition: false, phrase: null };
  }

  /**
   * 定型フレーズパターンを検出
   */
  detectPatternHallucination(text: string): boolean {
    if (!text) {
      return false;
    }

    for (const pattern of HALLUCINATION_PATTERNS) {
      if (pattern.test(text)) {
        return true;
      }
    }

    return false;
  }

  /**
   * テキストをフィルタリング
   */
  filter(text: string): HallucinationFilterResult {
    if (!this.enabled || !text) {
      return { text, wasFiltered: false, reason: null };
    }

    // パターンマッチによるハルシネーション検出
    if (this.detectPatternHallucination(text)) {
      this.stats.totalFiltered++;
      this.stats.patternFiltered++;
      logger.debug(`Pattern hallucination detected: ${text.substring(0, 50)}...`);
      return { text: '', wasFiltered: true, reason: 'pattern_match' };
    }

    // 繰り返し検出
    const { isRepetition, phrase } = this.detectRepetition(text);
    if (isRepetition) {
      this.stats.totalFiltered++;
      this.stats.repetitionFiltered++;
      logger.debug(
        `Repetition hallucination detected: '${phrase}' in '${text.substring(0, 50)}...'`
      );

      // 繰り返されたフレーズを1つだけ残す
      if (phrase) {
        return {
          text: phrase.trim(),
          wasFiltered: true,
          reason: `repetition:${phrase}`,
        };
      }

      return { text: '', wasFiltered: true, reason: 'repetition' };
    }

    return { text, wasFiltered: false, reason: null };
  }

  /**
   * 統計を取得
   */
  getStats(): {
    enabled: boolean;
    minRepetitionCount: number;
    maxRepetitionLength: number;
    totalFiltered: number;
    repetitionFiltered: number;
    patternFiltered: number;
  } {
    return {
      enabled: this.enabled,
      minRepetitionCount: this.minRepetitionCount,
      maxRepetitionLength: this.maxRepetitionLength,
      ...this.stats,
    };
  }

  /**
   * 統計をリセット
   */
  resetStats(): void {
    this.stats = {
      totalFiltered: 0,
      repetitionFiltered: 0,
      patternFiltered: 0,
    };
  }
}

File: whisper-api/src/services/hallucination_filter.py
"""Hallucination detection and filtering for Whisper transcription."""
import re
import logging
from typing import Tuple, Optional
from dataclasses import dataclass, field

logger = logging.getLogger(__name__)


@dataclass
class HallucinationStats:
    """Statistics for hallucination filtering."""
    
    total_filtered: int = 0
    repetition_filtered: int = 0
    pattern_filtered: int = 0
    
    def to_dict(self) -> dict:
        return {
            "total_filtered": self.total_filtered,
            "repetition_filtered": self.repetition_filtered,
            "pattern_filtered": self.pattern_filtered,
        }


class HallucinationFilter:
    """
    ハルシネーション（幻覚）フィルター
    
    Whisper はまれに以下のようなハルシネーションを生成する:
    - 同じフレーズの繰り返し: 「しょうがない しょうがない しょうがない」
    - 意味不明な繰り返しパターン
    - 特定の定型フレーズ（字幕提供、ご視聴ありがとうございましたなど）
    """

    # 定型フレーズパターン（Whisper がよく生成するハルシネーション）
    HALLUCINATION_PATTERNS = [
        r"^字幕提供.*$",
        r"^ご視聴ありがとうございました.*$",
        r"^チャンネル登録.*$",
        r"^お疲れ様でした$",
        r"^\.+$",  # ドットのみ
        r"^,+$",  # カンマのみ
        r"^[\s\u3000]+$",  # 空白のみ
        r"(?:music|♪|♫)+",  # 音楽記号
        r"\[音楽\]",
        r"\[拍手\]",
        r"^\s*お\s*$",  # 単独の「お」
        r"^\s*ん\s*$",  # 単独の「ん」
    ]

    def __init__(
        self,
        enabled: bool = True,
        min_repetition_count: int = 3,
        max_repetition_length: int = 20,
    ):
        """
        Initialize the hallucination filter.
        
        Args:
            enabled: Whether filtering is enabled.
            min_repetition_count: Minimum repetitions to detect as hallucination.
            max_repetition_length: Maximum length of repeated phrase to detect.
        """
        self.enabled = enabled
        self.min_repetition_count = min_repetition_count
        self.max_repetition_length = max_repetition_length
        self.stats = HallucinationStats()
        
        # Compile patterns
        self._compiled_patterns = [
            re.compile(p, re.IGNORECASE | re.UNICODE) 
            for p in self.HALLUCINATION_PATTERNS
        ]

    def detect_repetition(self, text: str) -> Tuple[bool, Optional[str]]:
        """
        繰り返しパターンを検出
        
        Args:
            text: Input text to check.
            
        Returns:
            Tuple of (is_repetition, repeated_phrase).
        """
        if not text or len(text) < 6:
            return False, None
        
        # 正規化
        text = text.strip()
        
        # スペースで分割して同じ単語の繰り返しを検出
        words = text.split()
        if len(words) >= self.min_repetition_count:
            # 全ての単語が同じかチェック
            unique_words = set(words)
            if len(unique_words) == 1:
                return True, words[0]
        
        # サブストリングの繰り返しを検出
        # 例: "abc abc abc" -> "abc" が繰り返し
        for phrase_len in range(2, min(self.max_repetition_length + 1, len(text) // 2 + 1)):
            phrase = text[:phrase_len]
            repetitions = text.count(phrase)
            
            # フレーズがテキストの大部分を占める場合
            if repetitions >= self.min_repetition_count:
                expected_len = len(phrase) * repetitions
                # 繰り返しがテキストの80%以上を占める
                if expected_len / len(text) >= 0.8:
                    return True, phrase
        
        return False, None

    def detect_pattern_hallucination(self, text: str) -> bool:
        """
        定型フレーズパターンを検出
        
        Args:
            text: Input text to check.
            
        Returns:
            True if text matches a hallucination pattern.
        """
        if not text:
            return False
        
        for pattern in self._compiled_patterns:
            if pattern.search(text):
                return True
        
        return False

    def filter(self, text: str) -> Tuple[str, bool, Optional[str]]:
        """
        テキストをフィルタリング
        
        Args:
            text: Input text to filter.
            
        Returns:
            Tuple of (filtered_text, was_filtered, reason).
        """
        if not self.enabled or not text:
            return text, False, None
        
        original_text = text
        
        # パターンマッチによるハルシネーション検出
        if self.detect_pattern_hallucination(text):
            self.stats.total_filtered += 1
            self.stats.pattern_filtered += 1
            logger.debug(f"Pattern hallucination detected: {text[:50]}...")
            return "", True, "pattern_match"
        
        # 繰り返し検出
        is_repetition, repeated_phrase = self.detect_repetition(text)
        if is_repetition:
            self.stats.total_filtered += 1
            self.stats.repetition_filtered += 1
            logger.debug(f"Repetition hallucination detected: '{repeated_phrase}' in '{text[:50]}...'")
            
            # 繰り返しを1回だけに減らす
            if repeated_phrase:
                # 繰り返されたフレーズを1つだけ残す
                cleaned = repeated_phrase.strip()
                return cleaned, True, f"repetition:{repeated_phrase}"
            
            return "", True, "repetition"
        
        return text, False, None

    def get_stats(self) -> dict:
        """Get filter statistics."""
        return {
            "enabled": self.enabled,
            "min_repetition_count": self.min_repetition_count,
            "max_repetition_length": self.max_repetition_length,
            **self.stats.to_dict(),
        }



